# Python optimization

In [1]:
import timeit
import requests

In [2]:
# Slow list append
def loop_list_append():
    list_object = []
    for i in range(100):
        list_object.append(i)


# Fast list append
def comprehension_list_append():
    [i for i in range(100)]


# Slow string concatenation
def addition_assignment_string():
    obj = ["hello", "my", "name", "is", "Ayub", "!"]
    s = ""
    for elem in obj:
        s += elem


# Fast string concatenation
def join_string():
    s = ["hello", "my", "name", "is", "Ayub", "!"]
    "".join(s)


# Slow loop append
def for_loop_append():
    obj = ["hello", "my", "name", "is", "Ayub", "!"]
    new = []
    for i in obj:
        new.append(i.upper())


# Fast map append
def get_upper(x):
    return x.upper()

def map_append():
    obj = ["hello", "my", "name", "is", "Ayub", "!"]
    map(get_upper, obj)


# Slow list membership
def create_word_iterable():
    word_site = "https://www.mit.edu/~ecprice/wordlist.10000"
    response = requests.get(word_site)
    word_list = response.content.splitlines()[0:50]  # 50 words
    word_set = set(word_list)
    return word_list, word_set

def list_membership():
    word_list, _ = create_word_iterable()
    if 'banana' in word_list:
        print("'banana' found in list")
    return


# Fast set membership
def set_membership():
    _, word_set = create_word_iterable()
    if 'banana' in word_set:
        print("'banana' found in set")
    return

### For loop vs. list comprehension

* Ran a for loop 100 times and appended to a list. Time was 4.181 seconds
* Ran a list comprehension 100 times. Time was 2.270 seconds.

In [3]:
t = timeit.Timer(setup='from __main__ import loop_list_append', stmt='loop_list_append()')
print(f"Time to run append 100 times = {t.timeit():.3f} seconds.")

Time to run append 100 times = 4.306 seconds.


In [4]:
t = timeit.Timer(setup='from __main__ import comprehension_list_append', stmt='comprehension_list_append()')
print(f"Time to run append 100 times = {t.timeit():.3f} seconds.")

Time to run append 100 times = 2.338 seconds.


### Avoid string concatenation

* Do not use += operator on strings
* Within a loop it is slow since strings are immutable
* Use `str.join()` method for efficient concatenation
* String concatenation using += took 0.380 seconds
* String concatenation using .join took 0.278 seconds

In [5]:
t = timeit.Timer(setup='from __main__ import addition_assignment_string', stmt='addition_assignment_string()')
print(f"Time to concatenate strings using += operator = {t.timeit():.3f} seconds.")

Time to concatenate strings using += operator = 0.424 seconds.


In [6]:
t = timeit.Timer(setup='from __main__ import join_string', stmt='join_string()')
print(f"Time to concatenate strings using .join operator = {t.timeit():.3f} seconds.")

Time to concatenate strings using .join operator = 0.208 seconds.


### For loop vs. Map

* Replace for loops with the more efficient `map()` function
* Apply a given function to each item of an iterable (list, tuple, string)
* Maps generate a new iterable containing the results of applying the function
* String concatenation using a for loop took 0.613 seconds
* String concatenation using map took 0.208 seconds

In [7]:
t = timeit.Timer(setup='from __main__ import for_loop_append', stmt='for_loop_append()')
print(f"Time to concatenate strings using a for loop = {t.timeit():.3f} seconds.")

Time to concatenate strings using a for loop = 0.664 seconds.


In [8]:
t = timeit.Timer(setup='from __main__ import map_append', stmt='map_append()')
print(f"Time to concatenate strings using map function = {t.timeit():.3f} seconds.")

Time to concatenate strings using map function = 0.231 seconds.


### List vs Set membership test

In [9]:
t = timeit.Timer(setup='from __main__ import list_membership', stmt='list_membership()')
print(f"Time to check if an element exists in a list = {t.timeit():.3f} seconds.")

In [ ]:
t = timeit.Timer(setup='from __main__ import set_membership', stmt='set_membership()')
print(f"Time to check if an element exists in a set = {t.timeit():.3f} seconds.")

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

KeyboardInterrupt: 